### PR TITLE
Refactor semantic3.d

### PR DESCRIPTION
**Pull request by thewilsonator**
Reduce variable scope and nesting. Use early returns and continues.

### CODE REVIEW

**Comment by thewilsonator** [comment]
Best reviewed with whitespace off.

File: src/dmd/semantic3.d [hunk 1]
```diff
@@ -1012,29 +1012,26 @@ private extern(C++) final class Semantic3Visitor : Visitor
                 auto a = new Statements();
                 // Merge in initialization of 'out' parameters
                 if (funcdecl.parameters)
+                foreach (v; *funcdecl.parameters)
```
Comment (WalterBright) [inline src/dmd/semantic3.d]: Not indenting after `if` is not an improvement.

File: src/dmd/semantic3.d [hunk 2]
```diff
@@ -1100,34 +1100,32 @@ private extern(C++) final class Semantic3Visitor : Visitor
                 /* Append destructor calls for parameters as finally blocks.
                  */
                 if (funcdecl.parameters)
+                foreach (v; *funcdecl.parameters)
```
Comment (WalterBright) [inline src/dmd/semantic3.d]: No on the not indent.

File: src/dmd/semantic3.d [hunk 3]
```diff
@@ -1289,15 +1289,14 @@ private extern(C++) final class Semantic3Visitor : Visitor
...
-                if (v.storage_class & STC.maybescope)
-                {
-                    //printf("Inferring scope for %s\n", v.toChars());
-                    Parameter p = f.parameterList[u];
-                    notMaybeScope(v);
-                    v.storage_class |= STC.scope_ | STC.scopeinferred;
-                    p.storageClass |= STC.scope_ | STC.scopeinferred;
-                    assert(!(p.storageClass & STC.maybescope));
-                }
+                if (!(v.storage_class & STC.maybescope))
```
Comment (WalterBright) [inline src/dmd/semantic3.d]: I'm not too thrilled about changing all the code base over things like this. For one thing, it adds a ! which is harder to cognitively process. Having straightforward positive expressions is better.

File: src/dmd/semantic3.d [hunk 4]
```diff
@@ -1428,17 +1428,17 @@ private extern(C++) final class Semantic3Visitor : Visitor
     override void visit(AttribDeclaration ad)
     {
         Dsymbols* d = ad.include(sc);
-        if (d)
+        if (!d)
+            return;
```
Comment (WalterBright) [inline src/dmd/semantic3.d]: Better to just use:
```
if (auto d = ad.include(sc))
```

**Comment by PetarKirov** [comment]
> ERROR: Error - file '/Users/braddr/sandbox/at-client/pull-3972371-Darwin_64_64/dmd/src/dmd/semantic3.d' contains trailing whitespace at line 1019

@thewilsonator Don't you have https://editorconfig.org/ set up?

File: src/dmd/semantic3.d (hunk 2 above)
Comment (PetarKirov) [reply src/dmd/semantic3.d]: I suggest the following:

```D
auto orEmpty(T)(inout Array!T* pArr)                          {
    return pArr? pArr.opSlice() : [];
}

void visit(FuncDeclaration funcdecl)                          {
    // ...
    foreach (v; funcdecl.parameters.orEmpty)
    {
    }
}
```

**Comment by UplinkCoder** [comment]
I don't particularly like the `orEmpty()` template you introduced.
If you didn't do that I would say the PR is an improvement.

File: src/dmd/semantic3.d (hunk 2 above)
Comment (WalterBright) [reply src/dmd/semantic3.d]: It's fine the way it was.

**Comment by RazvanN7** [comment]
This doesn't seem to be going anywhere. Refactorings should be smaller in size.

**Pull request closed without merging**